### PR TITLE
SILGen: Fix the logic of dynamic replacements for class constructors

### PR DESCRIPTION
**Pull request by aschwaighofer**
To correctly call designated super class initializers the designated
intializer (and not the allocator) is dynamically replaceable.
Convenience allocators are dynamically replaceable as before.

### CODE REVIEW

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Comment by aschwaighofer** [comment]
@swift-ci Please test

File: lib/IRGen/Linking.cpp [hunk 1]
```diff
@@ -317,8 +329,10 @@ std::string LinkEntity::mangleAsString() const {
     assert(isa<AbstractFunctionDecl>(getDecl()));
     std::string Result;
     if (auto *Constructor = dyn_cast<ConstructorDecl>(getDecl())) {
-      Result = mangler.mangleConstructorEntity(Constructor, true,
-                                               /*isCurried=*/false);
+      bool useAllocatorMangling = shouldUseAllocatorMangling(Constructor);
```
Comment (slavapestov) [inline lib/IRGen/Linking.cpp]: I would rather we had two separate LinkEntity kinds, DynamicallyReplaceableFunctionImpl and DynamicallyReplaceableAllocatorImpl. We already do this for method descriptors and dispatch thunks. Alternatively we could have a separate bit in LinkEntity for this purpose.

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Comment by aschwaighofer** [comment]
@swift-ci Please test os x platform

**Comment by aschwaighofer** [comment]
@swift-ci Please test os x platform

**Comment by aschwaighofer** [comment]
@swift-ci Please test os x platform

**Comment by aschwaighofer** [comment]
@swift-ci Please smoke test os x platform

**Comment by aschwaighofer** [comment]
Lets try this again with a rebased master

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Pull request merged**